File: src/main/java/de/blazemcworld/fireflow/code/node/impl/info/player/PlayerPositionNode.java
package de.blazemcworld.fireflow.code.node.impl.info.player;

import de.blazemcworld.fireflow.code.node.Node;
import de.blazemcworld.fireflow.code.type.PlayerType;
import de.blazemcworld.fireflow.code.type.PositionType;
import de.blazemcworld.fireflow.code.value.PlayerValue;
import net.minestom.server.coordinate.Pos;
import net.minestom.server.item.Material;

public class PlayerPositionNode extends Node {

    public PlayerPositionNode() {
        super("player_position", Material.RECOVERY_COMPASS);

        Input<PlayerValue> player = new Input<>("player", PlayerType.INSTANCE);
        Output<Pos> position = new Output<>("position", PositionType.INSTANCE);

        position.valueFrom((ctx) -> {
            PlayerValue p = player.getValue(ctx);
            if (p.available(ctx)) return p.get(ctx).getPosition();
            return Pos.ZERO;
        });
    }

    @Override
    public PlayerPositionNode copy() {
        return new PlayerPositionNode();
    }

}

File: src/main/kotlin/de/blazemcworld/fireflow/node/impl/player/info/PlayerPositionNode.kt
package de.blazemcworld.fireflow.node.impl.player.info

import de.blazemcworld.fireflow.node.BaseNode
import de.blazemcworld.fireflow.node.NodeContext
import de.blazemcworld.fireflow.node.PlayerType
import de.blazemcworld.fireflow.node.PositionType
import net.minestom.server.item.Material

object PlayerPositionNode : BaseNode("Player Position", Material.ENDER_EYE) {
    private val player = input("Player", PlayerType)
    private val position = output("Position", PositionType)

    override fun setup(ctx: NodeContext) {
        ctx[position].defaultHandler = { it[ctx[player]]?.resolve()?.position }
    }
}